public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx

```diff
@@ -88,6 +92,35 @@ export function panelMenuBehavior(menu: VizPanelMenu) {
   asyncFunc();
 }
 
+function addDataTrailAction(dashboard: DashboardScene, vizPanel: VizPanel, items: PanelMenuItem[]) {
+  const queryRunner = getQueryRunnerFor(vizPanel);
+  if (!queryRunner) {
+    return;
+  }
+
+  const ds = getDataSourceSrv().getInstanceSettings(queryRunner.state.datasource);
+  if (!ds || ds.meta.id !== 'prometheus' || queryRunner.state.queries.length > 1) {
+    return;
+  }
+
+  const query = queryRunner.state.queries[0];
+  const parsedResult = buildVisualQueryFromString(query.expr);
```

Comment:
I'd like us to make notes in the comments where we expect we could abstract certain things out, possibly to specific datasource implementations.

In this case, it looks like the ability to `buildVisualQueryFromString` could be what determines if we should exit out of the function or not (vs. checking meta id for `prometheus`). Would it be reasonable to add this as a method to the `DataSourceApi` interface? Then above, instead of checking if the datasource meta id is `prometheus`, we can `getDataSourceSrv().get(...)` and check if it has an implementation of `buildVisualQueryFromString`... or something similar. And then just use it.

I noticed that the Loki datasource code has a `buildVisualQueryFromString` function -- would it also be usable for this purpose?

I am not saying that we do this now, but if it's not much too early, perhaps we can add a point to do this sort of thing in:
- https://github.com/grafana/grafana/issues/78037

You mentioned there was another point you identified for potential abstraction. Can we add a comment in the code around that area too?

Reply:
yea, I think we can look at new abstractions some time next year as this becomes more solid / real